lib/dry/system/plugins/zeitwerk.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+require "dry/system/plugins/zeitwerk/compat_inflector"
+
+module Dry
+  module System
+    module Plugins
+      class Zeitwerk < Module
+        attr_reader :options
+
+        # @api private
+        def initialize(options)
+          @options = options
+          super()
+        end
+
+        # @api private
+        def extended(system)
+          system.before(:configure, &method(:configure_autoloading))
+          system.after(:configure, &method(:setup_autoloader))
+
+          super
+        end
+
+        private
+
+        # Configure autoloading on the container
+        #
+        # @return [self]
+        #
+        # @api private
+        def configure_autoloading(system)
+          require "dry/system/loader/autoloading"
+
+          system.setting :autoloader, reader: true
+          system.config.component_dirs.loader = Dry::System::Loader::Autoloading
+          system.config.component_dirs.add_to_load_path = false
+
+          system
+        end
+
+        # Set a logger
+        #
+        # This is invoked automatically when a container is being configured
+        #
+        # @return [self]
+        #
+        # @api private
+        def setup_autoloader(system)
+          return system if system.registered?(:autoloader)
+
+          if system.config.autoloader
+            system.register(:autoloader, system.config.autoloader)
+          else
+            system.config.autoloader = build_zeitwerk_loader(system)
+          end
```

Comment:
I'm a little confused by this... so if the user has provided their own autoloader instance via config, we register it in the container, but if we build our default one, we don't? What were you looking to achieve when you set this part up?

Would you actually want to register the autoloader in the container for both sides of that if/else clause? That said, I wonder if we actually want the autoloader in the container _at all_ — it feels too... _external_ to the application for us to want to have it readily accessible to all the application components themselves.

Reply:
Fair enough. I just copy-pasta'd this from the logging plugin. Happy to make it just register on the container instead